### PR TITLE
Update compilation 2020

### PR DESCRIPTION
**Pull request by quentinhocde**
- Remove gulp (using node scripts)
- Remove es5 support
- Add esbuild
- Rewrite tasks
- Keep mconfig.json from modularBP

### CODE REVIEW

**Comment by mcaskill** [comment]
Can something like ESLint be installed and applied to files to normalize our coding style?
There's quite a bit of superfluous blank lines, inconsistent indentation, missing white space (such as after commas), and other details which reduce readability.

**Comment by mcaskill** [comment]
Also, ensure the `.editorconfig` is in sync with the Charcoal boilerplate and that those rules are applied to this boilerplate.

**Review by Jerek0 — APPROVED** [review]

File: .editorconfig [hunk 1]
```diff
@@ -9,9 +9,12 @@ indent_style = space
 insert_final_newline = true
 trim_trailing_whitespace = true
 
-[*.md]
+[*.{md,markdown}]
 trim_trailing_whitespace = false
 
 [{*.yml,*.json}]
 indent_size = 4
 indent_style = space
```
Comment (mcaskill) [inline .editorconfig:17]:
```suggestion
```

This segment can be removed since it does not change from the initial rules.

File: README.md [hunk 2]
```diff
@@ -6,41 +6,40 @@
 <h1 align="center">Locomotive Boilerplate</h1>
 <p align="center">Front-end boilerplate for projects by Locomotive.</p>
 
+## Requirements
+
+| Name       | Version |
+| ---------- | ------- |
+| [Node]     | > 14.15  |
+[Node]:         https://nodejs.org/en/
```
Comment (mcaskill) [inline README.md:14]:
```suggestion
| Name       | Version  |
| ---------- | -------- |
| [Node]     | > 14.15  |

[Node]:         https://nodejs.org/en/
```

File: assets/scripts/modules/Load.js [hunk 3]
```diff
@@ -7,6 +7,7 @@ export default class extends module {
     }
 
     init() {
+
```
Comment (mcaskill) [inline assets/scripts/modules/Load.js:10]:
```suggestion
```

File: assets/scripts/modules/Scroll.js [hunk 4]
```diff
@@ -7,6 +7,7 @@ export default class extends module {
     }
 
     init() {
+
```
Comment (mcaskill) [inline assets/scripts/modules/Scroll.js:10]:
```suggestion
```

File: build/concat.js [hunk 5]
```diff
@@ -1,14 +1,34 @@
...
+        return true;
+      } else {
+        return false;
+      }
+    });
+
+    // Add absolute path
+    jsFiles = jsFiles.map((f) => {
+      return `${paths.scripts.vendors.src + f}`;
+    });
```
Comment (mcaskill) [inline build/concat.js:24]:
```suggestion
    // Exclude files that are not JavaScript
    var jsFiles = files.filter((file) => {
        return file.includes('.js');
    });

    // Prepend absolute path
    jsFiles = jsFiles.map((file) => {
        return `${paths.scripts.vendors.src + file}`;
    });
```

The `String.prototype.includes()` method returns a boolean, no need for `if else`.

File: build/concat.js [hunk 6]
```diff
@@ -1,14 +1,34 @@
...
+
+
```
Comment (mcaskill) [inline build/concat.js:26]:
```suggestion

```

File: build/concat.js [hunk 7]
```diff
@@ -1,14 +1,34 @@
...
+    // add files in node_modules example:
+    // jsFiles.push('node_modules/gsap/dist/gsap.min.js');
+
+    concat(jsFiles, paths.scripts.dest + paths.scripts.vendors.main + '.js').then(() => {
+        message('Vendors concatenated', 'success', 'Concat in');
+
```
Comment (mcaskill) [inline build/concat.js:32]:
```suggestion
```

File: build/concat.js [hunk 8]
```diff
@@ -1,14 +1,34 @@
...
+    })
+  }
```
Comment (mcaskill) [inline build/concat.js:34]:
```suggestion
}
```

File: build/notification.js [hunk 9]
```diff
@@ -0,0 +1,9 @@
+import notifier from 'node-notifier';
+
+export default function notification({title, message}) {
```
Comment (mcaskill) [inline build/notification.js:3]:
```suggestion
export default function notification({ title, message }) {
```

File: build/scripts.js [hunk 10]
```diff
@@ -0,0 +1,31 @@
+import esbuild from 'esbuild';
+import paths from '../mconfig.json';
+import message from './utils/message.js';
+import notification from './notification.js';
+
+export function buildScripts() {
+    console.time('JS built in');
+
+    esbuild.build({
+        entryPoints: [paths.scripts.src + paths.scripts.main + '.js'],
```
Comment (mcaskill) [inline build/scripts.js:10]:
```suggestion
        entryPoints: [ paths.scripts.src + paths.scripts.main + '.js' ],
```

File: build/scripts.js [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+        bundle: true,
+        minify: true,
+        sourcemap: true,
+        color: true,
+        logLevel: 'error',
+        errorLimit: 1,
+        target: ['es2015'],
```
Comment (mcaskill) [inline build/scripts.js:17]:
```suggestion
        target: [ 'es2015' ],
```

File: build/scripts.js [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+
+        notification({
+            title: 'Javascript built failed 🚨',
+            message: `${e.errors[0].text} in ${e.errors[0].location.file} line ${e.errors[0].location.line}`
+        });
+
+    }).then(() => {
+        message('Javascript built','success', 'JS built in')
+    })
+
```
Comment (mcaskill) [inline build/scripts.js:30]:
```suggestion
```

File: build/styles.js [hunk 13]
```diff
@@ -0,0 +1,59 @@
+
```
Comment (mcaskill) [inline build/styles.js:1]:
```suggestion
```

File: build/styles.js [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+    // Compile critical scss
+    sass.render({
+        file: paths.styles.src + paths.styles.critical + '.scss',
+        outFile: paths.styles.dest + paths.styles.critical + '.css',
+        outputStyle: 'compressed',
+        sourceMap: true
+    }, (error, result) => {
+
+        if(error) {
```
Comment (mcaskill) [inline build/styles.js:47]:
```suggestion
        if (error) {
```

File: build/styles.js [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+
+    // Compile main scss
+    sass.render({
+        file: paths.styles.src + paths.styles.main + '.scss',
+        outFile: paths.styles.dest + paths.styles.main + '.css',
+        outputStyle: 'compressed',
+        sourceMap: true
+    }, (error, result) => {
+
+        if(error) {
```
Comment (mcaskill) [inline build/styles.js:19]:
```suggestion
        if (error) {
```

File: build/styles.js [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+                message: `${error.formatted}`
+            });
+        } else {
+            message('Styles built', 'success', 'Styles built in');
+        }
+
+        if(!error){
+            // No errors during the compilation, write this result on the disk
+            fs.writeFile(paths.styles.dest + paths.styles.main + '.css', result.css, (err) => {});
+          }
```
Comment (mcaskill) [inline build/styles.js:34]:
```suggestion
        if (!error){
            // No errors during the compilation, write this result on the disk
            fs.writeFile(paths.styles.dest + paths.styles.main + '.css', result.css, (err) => {});
        }
```

File: build/styles.js [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
+            message('Error compiling critical.scss', 'error');
+            console.log(error);
+        } else {
+            message('Critical style built', 'success', 'Critical style built in');
+        }
+
+        if(!error){
+            // No errors during the compilation, write this result on the disk
+            fs.writeFile(paths.styles.dest + paths.styles.critical + '.css', result.css, (err) => {});
+          }
```
Comment (mcaskill) [inline build/styles.js:57]:
```suggestion
        if (!error){
            // No errors during the compilation, write this result on the disk
            fs.writeFile(paths.styles.dest + paths.styles.critical + '.css', result.css, (err) => {});
        }
```

File: build/svgs.js [hunk 18]
```diff
@@ -0,0 +1,16 @@
+import mixer from 'svg-mixer';
+import paths from '../mconfig.json';
+import message from './utils/message.js';
+
+export function generateSpriteSVG() {
+
```
Comment (mcaskill) [inline build/svgs.js:6]:
```suggestion
```

File: build/svgs.js [hunk 19]
```diff
@@ -0,0 +1,16 @@
...
+    console.time('Sprite generated in');
+
+    // Write sprite content on disk
+    mixer([paths.svgs.src + '*.svg'], {spriteConfig: { usages: false }})
+    .then((result) => {
+        result.write(paths.svgs.dest + 'sprite.svg');
+        message('SVG Sprite generated', 'success', 'Sprite generated in');
+
+    });
```
Comment (mcaskill) [inline build/svgs.js:15]:
```suggestion
    mixer([
        paths.svgs.src + '*.svg'
    ], {
        spriteConfig: {
            usages: false
        }
    }).then((result) => {
        result.write(paths.svgs.dest + 'sprite.svg');
        message('SVG Sprite generated', 'success', 'Sprite generated in');
    });
```

File: build/utils/message.js [hunk 20]
```diff
@@ -0,0 +1,23 @@
+
```
Comment (mcaskill) [inline build/utils/message.js:1]:
```suggestion
```

File: build/utils/message.js [hunk 21]
```diff
@@ -0,0 +1,23 @@
...
+        }
+    } else if (type === 'error') {
+        console.log(kleur.red().underline(`${text} ❌`));
+
+    } else if (type === 'waiting') {
+
+        console.log(kleur.blue().italic(`${text} ⏱ `));
+    } else {
+        console.log(text);
+    }
```
Comment (mcaskill) [inline build/utils/message.js:20]:
```suggestion
    if (type === 'success') {
        console.log(kleur.bgGreen().black(`${text} ✅`));

        if (timerId !== undefined) {
            console.timeEnd(timerId)
        }
    } else if (type === 'error') {
        console.log(kleur.red().underline(`${text} ❌`));
    } else if (type === 'waiting') {
        console.log(kleur.blue().italic(`${text} ⏱`));
    } else {
        console.log(text);
    }
```

File: package.json [hunk 22]
```diff
@@ -3,20 +3,28 @@
   "name": "@locomotivemtl/boilerplate",
   "title": "Locomotive Boilerplate",
   "version": "1.0.0",
+  "engines": { "node": "???" },
```
Comment (mcaskill) [inline package.json:6]:
```suggestion
  "engines": {
    "node": "^14.15"
  },
```

This should reflect the version defined in the `.nvmrc`.

**Review by mcaskill — CHANGES_REQUESTED** [review]
#### package.json, package-lock.json, mconfig.json

Files should adhere to a 4-space indentation as per `.editorconfig`.

#### assets/scripts/app.js

The [app.js](https://github.com/locomotivemtl/locomotive-boilerplate/blob/update-compilation-2020/assets/scripts/app.js) contains inconsistent string quotation characters.

```diff
-    const $style = document.getElementById("stylesheet");
+    const $style = document.getElementById('stylesheet');
```

This ID should maybe be dynamic or reflect the name of the stylesheet file name, similar to what WordPress does for its enqueued styles.

```diff
-    const $style = document.getElementById('stylesheet');
+    const $style = document.getElementById('main-css');
```

#### assets/scripts/modules/Scroll.js

The [Scroll.js](https://github.com/locomotivemtl/locomotive-boilerplate/blob/update-compilation-2020/assets/scripts/modules/Scroll.js) contains inconsistent coding styles.

```diff
-            if(args.obj.el.tagName == "IMG") {
-                args.obj.el.src = src
-            } else {
-                args.obj.el.style.backgroundImage = `url(${src})`
-            }
+            if (args.obj.el.tagName === 'IMG') {
+                args.obj.el.src = src
+            } else {
+                args.obj.el.style.backgroundImage = `url('${src}')`
+            }
```

**Comment by quentinhocde** [comment]
Thanks @mcaskill, I fix everything in your review :)

**Review by mcaskill — APPROVED** [review]

**Review by devenini — APPROVED** [review]

**Review by arnvvd — APPROVED** [review]

**Pull request merged**